src/Compilers/VisualBasic/Portable/Symbols/Source/SourceFile.vb

```diff
@@ -331,29 +310,67 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' Specifically, constraints are checked for generic type references.
         ''' </summary>
-        Private Shared Sub ValidateImports(memberImports As ImmutableArray(Of NamespaceOrTypeAndImportsClausePosition),
+        Private Shared Sub ValidateImports(compilation As VisualBasicCompilation,
+                                           memberImports As ImmutableArray(Of NamespaceOrTypeAndImportsClausePosition),
                                            memberImportsSyntax As ImmutableArray(Of SyntaxReference),
                                            aliasImportsOpt As IReadOnlyDictionary(Of String, AliasAndImportsClausePosition),
                                            diagnostics As BindingDiagnosticBag)
             ' TODO: Dev10 reports error on specific type parts rather than the import
             ' (reporting error on Object rather than C in C = A(Of Object) for instance).
+            Dim clauseDiagnostics = BindingDiagnosticBag.GetInstance()
 
             For i = 0 To memberImports.Length - 1
-                Dim type = TryCast(memberImports(i).NamespaceOrType, TypeSymbol)
-                If type IsNot Nothing Then
-                    Dim location = memberImportsSyntax(i).GetLocation()
-                    type.CheckAllConstraints(location, diagnostics)
-                End If
+                ValidateImportsClause(compilation, clauseDiagnostics, memberImports(i).NamespaceOrType,
+                                      memberImports(i).Dependencies, memberImportsSyntax(i).GetLocation(),
+                                      memberImports(i).ImportsClausePosition, diagnostics)
             Next
 
             If aliasImportsOpt IsNot Nothing Then
                 For Each aliasImport In aliasImportsOpt.Values
-                    Dim type = TryCast(aliasImport.Alias.Target, TypeSymbol)
-                    If type IsNot Nothing Then
-                        type.CheckAllConstraints(aliasImport.Alias.Locations(0), diagnostics)
-                    End If
+                    ValidateImportsClause(compilation, clauseDiagnostics, aliasImport.Alias.Target,
+                                          aliasImport.Dependencies, aliasImport.Alias.Locations(0),
+                                          aliasImport.ImportsClausePosition, diagnostics)
                 Next
             End If
+
+            clauseDiagnostics.Free()
+        End Sub
+
+        Private Shared Sub ValidateImportsClause(
+            compilation As VisualBasicCompilation,
+            clauseDiagnostics As BindingDiagnosticBag,
+            namespaceOrType As NamespaceOrTypeSymbol,
+            dependencies As ImmutableArray(Of AssemblySymbol),
+            location As Location,
+            importsClausePosition As Integer,
+            diagnostics As BindingDiagnosticBag
+        )
+            Dim type = TryCast(namespaceOrType, TypeSymbol)
+            If type IsNot Nothing Then
+                clauseDiagnostics.Clear()
```

Comment:
I think it would be better to clear these on every loop in the calling method, rather than splitting the semantics of how this particular `BindingDiagnosticsBag` is supposed to be used across multiple methods.

Reply:
> I think it would be better to clear these on every loop in the calling method, rather than splitting the semantics of how this particular BindingDiagnosticsBag is supposed to be used across multiple methods.

I don't think I agree. The caller is only responsible for providing an instance, it doesn't use this bag at all. Usage in this method depends on starting with an empty bag, therefore, I believe it is more appropriate to do this here, rather than adding an extra contract for the caller. That, in fact would split usage semantics across multiple methods, I think.

---
In reply to: [369268016](https://github.com/dotnet/roslyn/pull/40987#discussion_r369268016) [](ancestors = 369268016)